File: datasets/python_source-code/Multiple-C2/organic_4.py
"""
0/1 Knapsack (DP) - Organic Submission #4
Top-down recursion using functools.lru_cache instead of a manual dict.
"""

from functools import lru_cache


def knapsack(weights, values, capacity):
    weights = tuple(weights)
    values = tuple(values)
    n = len(weights)

    @lru_cache(maxsize=None)
    def helper(i, cap):
        if i == n or cap == 0:
            return 0
        without_item = helper(i + 1, cap)
        if weights[i] > cap:
            return without_item
        with_item = values[i] + helper(i + 1, cap - weights[i])
        return max(without_item, with_item)

    result = helper(0, capacity)
    helper.cache_clear()
    return result


if __name__ == "__main__":
    weights = [1, 2, 3]
    values = [6, 10, 12]
    capacity = 5
    print(knapsack(weights, values, capacity))

File: server/datasets/controlled_benchmarks/java/multiple_files/multiple_c2/organic_4.java
/*
 * 0/1 Knapsack (DP) - Organic Submission #4
 * Top-down recursion with 2D memoization array cache.
 */

class Solution {
    private static int helper(int i, int cap, int[] weights, int[] values, Integer[][] memo) {
        if (i == weights.length || cap == 0) {
            return 0;
        }
        if (memo[i][cap] != null) {
            return memo[i][cap];
        }

        int withoutItem = helper(i + 1, cap, weights, values, memo);
        if (weights[i] > cap) {
            memo[i][cap] = withoutItem;
            return withoutItem;
        }

        int withItem = values[i] + helper(i + 1, cap - weights[i], weights, values, memo);
        int result = Math.max(withoutItem, withItem);
        memo[i][cap] = result;
        return result;
    }

    public static int knapsack(int[] weights, int[] values, int capacity) {
        Integer[][] memo = new Integer[weights.length + 1][capacity + 1];
        return helper(0, capacity, weights, values, memo);
    }

    public static void main(String[] args) {
        int[] weights = {1, 2, 3};
        int[] values = {6, 10, 12};
        int capacity = 5;
        System.out.println(knapsack(weights, values, capacity));
    }
}
